### PR TITLE
Fix dlts_get_max_record_overhead()

### PR DESCRIPTION
**Pull request by mattcaswell**
The calculation of the maximum DTLS record overhead was incorrect which means that we could exceed the MTU in some situations. We fix the calculation to take account of more sources of record overhead. We ignore potential overheads due to expansion after compression. At the time we make the calculation we don't know how long the record will be yet. However we can't accurately calculate the expansion without knowing that length. Just using a fallback expansion value for the worst case scenario doesn't help because that value is too high and is bigger than our fallback MTU size.

Testing this was tricky. The best solution I've come up with is to add an assert into statem_dtls.c to verify that we never exceed the MTU. Without the fix in dtls_get_max_record_overhead() this causes the existing test suite to fail. With the fix the test suite still passes.

While investigating this problem I spotted and fixed a different but related problem in tls_common.c. We calculates the maximum amount of growth we might expect to see during an encryption operation and reserve that many bytes in the WPACKET. Unfortunately the calculation was wrong which means that the encryption can overflow the amount of bytes reserved for it. In practice this isn't really an issue because the underlying buffer is still sufficiently large.

### CODE REVIEW

**Review by hlandau — APPROVED** [review]

**Comment by mattcaswell** [comment]
Ping for second review?

**Review by paulidale — APPROVED** [review]
I'm sure I'd already reviewed this :confused:

**Comment by mattcaswell** [comment]
Oops. I might have prematurely merged this. CI failure looks potentially relevant. Investigating...

**Comment by mattcaswell** [comment]
Why did openssl-machine add "ready to merge" when there was a CI failure??

**Comment by t8m** [comment]
> Why did openssl-machine add "ready to merge" when there was a CI failure??

That's interesting. I've never seen such occurrence before.

**Comment by mattcaswell** [comment]
Fix in #19622

**Pull request closed without merging**
